File: src/Cryptofolio.Infrastructure/Entities/Transaction.cs
using System;
using System.Text.Json.Serialization;

namespace Cryptofolio.Infrastructure.Entities
{
    /// <summary>
    /// Models a transaction.
    /// </summary>
    public abstract class Transaction
    {
        /// <summary>
        /// The id.
        /// </summary>
        public string Id { get; set; }

        /// <summary>
        /// The date when it occured.
        /// </summary>
        public DateTimeOffset Date { get; set; }

        /// <summary>
        /// The wallet it's attached to.
        /// </summary>
        public Wallet Wallet { get; set; }

        /// <summary>
        /// The asset it concerns.
        /// </summary>
        public Asset Asset { get; set; }

        /// <summary>
        /// The exchange on which it was bought/sold/transfered.
        /// </summary>
        public Exchange Exchange { get; set; }

        /// <summary>
        /// The asset quantity.
        /// </summary>
        public decimal Qty { get; set; }

        /// <summary>
        /// The initial value (computed with the ticker at the transaction date) of the transaction.
        /// </summary>
        [JsonPropertyName("initial_value")]
        public decimal InitialValue { get; set; }

        /// <summary>
        /// The current value (computed with the latest ticker) of the transaction.
        /// </summary>
        [JsonPropertyName("current_value")]
        public decimal CurrentValue { get; set; }

        /// <summary>
        /// The change in percent.
        /// </summary>
        public decimal Change { get; set; }

        /// <summary>
        /// The optional note.
        /// </summary>
        public string Note { get; set; }

        /// <summary>
        /// The type discriminator.
        /// </summary>
        [JsonPropertyName("type_discriminator")]
        public string TypeDiscriminator => GetType().Name;
    }
}

File: src/Cryptofolio.App/ClientApp/src/app/models/transaction.ts
import { Asset } from "./asset";
import { Currency } from "./currency";
import { Exchange } from "./exchange";
import { Wallet } from "./wallet";

export abstract class Transaction {
  static readonly BuyOrSellType = "BuyOrSellTransaction";
  static readonly TransferType = "TransferTransaction";

  id: string;
  date: Date;
  wallet: Wallet;
  asset: Asset;
  exchange: Exchange;
  qty: number;
  initialValue: number;
  currentValue: number;
  change: number;
  note: string;
  type_discriminator: string;
}

export class BuyOrSellTransaction extends Transaction {
  type: string;
  currency: Currency;
  price: number;

  constructor(properties?: any) {
    super();
    Object.assign(this, properties);
    this.date = new Date(properties.date);
    this.wallet = new Wallet(properties.wallet);
    this.asset = new Asset(properties.asset);
    this.exchange = new Exchange(properties.exchange);
    this.currency = new Currency(properties.currency);
    this.initialValue = properties["initial_value"];
    this.currentValue = properties["current_value"];
  }
}

export class TransferTransaction extends Transaction {
  source: string;
  destination: string;

  constructor(properties?: any) {
    super();
    Object.assign(this, properties);
    this.wallet = new Wallet(properties.wallet);
    this.asset = new Asset(properties.asset);
    this.exchange = new Exchange(properties.exchange);
    this.initialValue = properties["initial_value"];
    this.currentValue = properties["current_value"];
  }
}
